### ISSUE TITLE
Rules: Condition with attribute with RULES_RESET_IMMEDIATE reset immediately

### ISSUE THREAD

**Issue by DonWillems** [openremote/openremote#434]
As a user I want to create rules for events that trigger on every update, so that I can send notifications for every alarm that comes in on an attribute.

- [x] Add configuration item for attributes: RULES_RESET_IMMEDIATE
- [ ] A (When) condition with an attribute with that config item resets immediately after each value update.

**Comment by pierrekil** [comment]
For LHS condition the rule doesn't retrigger when there are two conditions (AND) of which one receives a new value and includes RULES_RESET_IMMEDIATE and the other doesn't update and remains TRUE:
![Screenshot 2021-06-08 at 12 50 41](https://user-images.githubusercontent.com/11444149/121172994-92c51980-c858-11eb-92d6-e4c693bd2730.png)

**Comment by richturner** [comment]
If I'm reading your comment correctly then this behaviour sounds correct as it is an AND condition. Each attribute in the condition will need the `RULES_RESET_IMMEDIATE` configuration item.

**Comment by pierrekil** [comment]
I have added it to 'Area' attribute, but still no repeat. Is it because the 'area' did not receive a new value (it's static)?

**Comment by richturner** [comment]
Each condition needs some sort of latch otherwise it would continue to fire each time the rule is evaluated. Even with `RULES_RESET_IMMEDIATE` a new value must be received for the attribute in order to reset it.

Maybe we need to revisit this logic and potentially reset all conditions once the rule fails to match

**Comment by pierrekil** [comment]
What about:
Reset LHS when any of the attributes on LHS has a RulesResetImmediately and evaluate again only when at least one of the LHS attributes receives new data

**Comment by richturner** [comment]
Need a more generalised approach not just for `RULES_RESET_IMMEDIATE`, i.e.

WHEN Light ON and Outside Brightness > 1000lux
THEN SEND ALERT `DON'T WASTE ENERGY`

If the light goes on then the rule triggers, if the light then goes off and comes back on whilst Brightness stays above 1000lux then the alert should be sent again but currently it won't until both conditions stop matching.

**Comment by pierrekil** [comment]
So, does "reset all conditions once the rule fails to match" also cover a reset caused by the RULES_RESET_IMMEDIATE? In that case it makes a lot of sense to do this to cover a more generic behavior

**Comment by richturner** [comment]
It is more complicated than this because a rule could apply to multiple assets

**Comment by aktur** [comment]
In boolean algebra AND truth table is:
ABo
000
010
100
111

Please don't implement other type of algebra, because the boolean one is proved to work, which computers are a good example off. In the example you show you've implemented a latch with following truth table:
ABo
000
01u
10u
111
Where 'p' denotes previous state which is undefined. If you are wanting to implement memory element please do register instead of latch. A common way of doing this is putting a clock to sync states. More about this:

https://electronics.stackexchange.com/questions/38645/why-are-inferred-latches-bad#38650
https://electricalvoice.com/difference-between-latch-and-flip-flop/

In short, latches should be avoided and flip-flip flops should be used. Some basic logic circuit design textbook should be helpful, e.g. https://my.ece.utah.edu/~kalla/ECE3700/verlogic3_chapter2.pdf

**Comment by richturner** [comment]
We need to use latching to avoid infinite loops.

With attributes of continuous data type (double), you don't want a greater than Y rule to continue triggering for each value above Y, the common sense thing to do here is to latch until the condition no longer matches which is indeed implicit in the when then rules engine but multiple conditions start to make this behave in a non-intuitive way.

I would be interested in how this behaviour works in drools for example, if the when includes multiple facts when is the rule eligible for re-evaluation, any ideas @aktur?

When then rules have the complexity of allowing a condition to be scoped to any asset of a type so you could have a rule like:

WHEN Any RoomAsset Light ON and Outside Brightness > 1000lux
THEN SEND ALERT DON'T WASTE ENERGY

Any latching logic needs to be appropriately scoped.

**Comment by aktur** [comment]
For preventing infinitive loops you need to add delay not latching. In Drools simply by adding (int: 1) which you can spot in many rules I've written.

For continuous data you use inference (i.e. one rule triggers another one). In your example it would be (this is pseudocode, as you would define custom facts types for this):

rule a
  when
    Y > 1000
    not isAbove1000
then
  insert(new isAbove1000)
end

rule b
  when
    isAbove1000
then
  //trigger alarm
end

rule c
  when
   Y <= 1000
  $p: isAbove1000
then
  retract($p)
end

This way the rule b will trigger only once when Y cross above 1000. This is for the old Drools used in 2.5, in newer versions they introduced syntax to shorthand this pattern to a single LHS, so you don't need inference anymore.

With multiply facts you use similar pattern, a very good example is my Green StrijpS/Block 61 Drools example. There I've implemented this inference with multiply facts in few places. However, this was replaced by some guy who did not fully understood what I've done there. The rules, which would never work without it (occupancy self-learning prediction) he just removed from the loop and it was done. Anyway, I think that with the newest Drools it would be just a single rule, but I've drop my Drools interest two years ago so don't know about the newest developments. The only thing that I'm aware it is now possible to deploy Drools as Quarkus/GraalVM rest serverless service 
https://developers.redhat.com/blog/2019/03/14/quarking-drools-how-we-turned-a-13-year-old-java-project-into-a-first-class-serverless-component

**Comment by aktur** [comment]
```
WHEN Any RoomAsset Light ON and Outside Brightness > 1000lux
THEN SEND ALERT DON'T WASTE ENERGY
```
In the old Drools for 2.5 I would write something like this:
```
declare Alarm
  name: String
end

rule a
when
  exists Event(asset=='RoomAsset',  value=='Light ON')
  Event(asset=='Outside Brightness', value > 1000lux)
  not Alarm(name=='rule a alarm')
then
  insert(new Alarm('rule a alarm'));
  SEND_ALERT("DON'T WASTE ENERGY");
end

rule a1
when
  Event(asset=='RoomAsset',  value!='Light ON')
  $a: Alarm(name=='rule a alarm')
then
  retract($a);
end

rule a2
when
  Event(asset=='Outside Brightness', value <= 1000lux)
  $a: Alarm(name=='rule a alarm')
then
  retract($a);
end
```
This should be a little bit closer to correct syntax. So you see one rule to set the alarm and 2 rules to remove the alarm condition. Those 2 rules could be combined to a single rule with OR on the LHS, but OR on the LHS is discouraged in Drools and I prefer to write separate rule for each part of the OR. This is how a hardware would be synthesised and from my background it is more natural. A software engineer would prefer single rule in order to gain on not copying the retract($a) statement. Anyway, both version would work. 

In the new  Drools they've introduce a DSL (Domain Specific Language) and perhaps it would be possible to create a DSL which would parse correctly your original rule 

WHEN Any RoomAsset Light ON and Outside Brightness > 1000lux
THEN SEND ALERT DON'T WASTE ENERGY

**Comment by aktur** [comment]
Alternatively, rule a1 could be
```
rule a1
when
  not exists Event(asset=='RoomAsset',  value=='Light ON')
  $a: Alarm(name=='rule a alarm')
then
  retract($a);
end
```
Depending on what your intention is. Perhaps in lights situation this would be a better choice. In FIRE alarm situation I would prefer the previous one.

**Comment by richturner** [comment]
Using these temporary facts (state) is latching and I do this implicitly in our JSON (when then) rules engine to hide this complexity from target users of when then rules.

Drools does some instance checking on the fact that caused the match to prevent the rule from firing again (when `modify` is used to alter the fact which obviously mutates the fact and makes it eligible for matching again). I do something similar in our when then rules which introduces the problem of interest here "When does the rule become eligible for firing again".

In drools I cannot remember what the behaviour is when you use multiple facts in the LHS (AND logic), the rule then matches and only one of the facts in the LHS is mutated and the LHS still evaluates to true, would the rule fire again or would all LHS facts need to be mutated to allow the rule to fire again.

**Comment by aktur** [comment]
Default in Drools a rule matches when all facts evaluate true (AND behaviour) on each fact update. The second part was a problem in controller 2.5 therefore there was Java preprocessor which updated a fact only when associated sensor changed value (but this is not Drools behaviour). So on sensor change you had fact retraction and insertion. Ch. removed this as 'stupid' and hence you had multiple rules triggering without real change. For example, in Schiphol many rules were re-triggered on each new API pool. I've solved this by insertion and retraction in Drools, so in fact redoing what 2.5 preprocessor did but now pure in Drools. Again Ch. said that this is stupid and replaced Drools with Groovy. Now, you have to redo everything which Drools implemented 13 years of development from scratch. As you see, it is not so easy.

As your question about firing rules, in Drools you have everything under control. You can define any triggering you wish. In fact you can model any hardware, even make a hardware simulator in Drools. On the other hand, in hardware world when you synthesise latch instead of flip-flop then you get an error, as this hardware would usually break in the real world. I would not go into discussing digital circuits stability and timing analysis as these are quite complex subjects. Drools implements it correctly. Now if you want to custom implement it then I would advise to start with reading some digital design textbook and learn why latches should be avoided. Maybe you can hire some digital designer who can help you? Or you can replace custom build rules engine with some industry proven one? Take Amadeus, they have many digital PhD and yet I was asked to help them in replacing custom build rule engine, written in C++, with Drools. Their C++ engine was correct, but the problem with it is that C++ designers are dying out and they have bigger and bigger problem with new hires to support their custom C++. Hence they chosen one supported by external company (RedHat) but hit another wall. As Drools is Java based it uses garbage collector which is very problematic in real time processing causing unpredictable delays. But these are problems at scale of Amadeus, when a rule engine has millions hits per second. We don't have this problem yet so Drools would be a good choice.

**Comment by aktur** [comment]
_Using these temporary facts (state) is latching and I do this implicitly in our JSON (when then) rules engine to hide this complexity from target users of when then rules._

It is not called latching but _using a memory element_. A latch is an async element while flip-flop is usually master-slave element consisting from at least 2 latches. There are few types of flip-flops like D, T, JK and RS. For flexibility you would like to have all of them available. From what I understand you just implemented a single latch. My Drools rule is an example of RS flip-flop = separate inputs for Set and Reset.

**Comment by pierrekil** [comment]
Another real case, registered on forum: https://forum.openremote.io/t/rule-triggered-by-timer-not-implemented/872/5
